### ISSUE TITLE
[BUG]: Cannot restore the lr_scheduler correctly

### ISSUE THREAD

**Issue by ericxsun** [hpcaitech/ColossalAI#5151]
### 🐛 Describe the bug

After saving optimizer and learning rate scheduler states after 6,000 steps, and successfully restore the states , I observe an incorrect learning rate at the first step after restoration (step 6,001). Specifically, the learning rate(`lr_scheduler.get_last_lr()[0]`  or `lr_scheduler.get_lr()[0]`) seems to be reset to the initial value.

![image](https://github.com/hpcaitech/ColossalAI/assets/1772912/49ec74d1-ddcb-4bac-830e-03300c5acb5d)

```
plugin=gemini
booster = Booster(plugin=plugin)
....
optimizer = HybridAdam(model.parameters(), lr=args.lr, betas=(0.9, 0.95), weight_decay=args.weight_decay)

lr_scheduler = CosineAnnealingWarmupLR(
    optimizer, total_steps=args.max_steps, warmup_steps=args.warmup_steps, eta_min=0.1 * args.lr
)
....

for step in dataloader:
    ....
    optimizer.step()
    lr_scheduler.step()
    optimizer.zero_grad()
    ....
    log.info('lr', lr_scheduler.get_last_lr()[0])
```

### Environment

Pytorch: 2.0.1
ColossalAI: 0.3.4

**Comment by imgaojun** [comment]
It appears that the issue you're encountering is due to the `param_group['lr']` not being saved correctly. This is a common challenge when restoring the state of an optimizer and learning rate scheduler in deep learning models. As of now, I haven't been able to determine the exact reason why `param_group['lr']` is not saved as expected. However, to address this, I've developed a temporary solution that ensures the learning rate is correctly set upon restoration. Here's the refined code snippet:

```python
def load_checkpoint(booster, model, optimizer, lr_scheduler):
    args = get_args()
    load_dir = args.load_dir
    
    # Load model, optimizer, and lr_scheduler states
    booster.load_model(model, os.path.join(load_dir, "model"))
    booster.load_optimizer(optimizer, os.path.join(load_dir, "optimizer"))
    booster.load_lr_scheduler(lr_scheduler, os.path.join(load_dir, "lr_scheduler"))

    # Correctly set the learning rate for each param_group
    for i, data in enumerate(zip(optimizer.param_groups, lr_scheduler.get_last_lr())):
        param_group, lr = data
        param_group['lr'] = lr
```

This code ensures that each parameter group's learning rate is updated to match the last recorded learning rate from the scheduler. By doing this, we can maintain consistency in the learning rate when resuming training from a checkpoint. This approach is a workaround until a more permanent solution is identified for the underlying issue of `param_group['lr']` not being saved properly.

**Comment by ericxsun** [comment]
> It appears that the issue you're encountering is due to the `param_group['lr']` not being saved correctly. As of now, I haven't been able to determine the exact reason why `param_group['lr']` is not saved as expected. However, to address this, I've developed a temporary solution that ensures the learning rate is correctly set upon restoration. Here's the refined code snippet:
> 
> ```python
> def load_checkpoint(booster, model, optimizer, lr_scheduler):
>     args = get_args()
>     load_dir = args.load_dir
>     
>     # Load model, optimizer, and lr_scheduler states
>     booster.load_model(model, os.path.join(load_dir, "model"))
>     booster.load_optimizer(optimizer, os.path.join(load_dir, "optimizer"))
>     booster.load_lr_scheduler(lr_scheduler, os.path.join(load_dir, "lr_scheduler"))
> 
>     # Correctly set the learning rate for each param_group
>     for i, data in enumerate(zip(optimizer.param_groups, lr_scheduler.get_last_lr())):
>         param_group, lr = data
>         param_group['lr'] = lr
> ```
> 
> This code ensures that each parameter group's learning rate is updated to match the last recorded learning rate from the scheduler. By doing this, we can maintain consistency in the learning rate when resuming training from a checkpoint. This approach is a workaround until a more permanent solution is identified for the underlying issue of `param_group['lr']` not being saved properly.

Thanks so much @imgaojun

**Comment by ericxsun** [comment]
Perhaps using `lr_scheduler.step(step)` after restoring the learning rate scheduler is a simpler approach, if we know the step when saving lr_scheduler.
